File: modules/react-native-status/android/src/main/java/im/status/ethereum/module/DatabaseManager.java
package im.status.ethereum.module;

import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.Callback;
import statusgo.Statusgo;
import android.util.Log;
import java.io.File;
import android.os.Environment;
import android.content.Context;
public class DatabaseManager extends ReactContextBaseJavaModule {
    private static final String TAG = "DatabaseManager";
    private ReactApplicationContext reactContext;
    private static final String exportDBFileName = "export.db";
    private Utils utils;
    public DatabaseManager(ReactApplicationContext reactContext) {
        this.reactContext = reactContext;
        this.utils = new Utils(reactContext);
    }
    @Override
    public String getName() {
        return "DatabaseManager";
    }

    private File getExportDBFile() {
        final Context context = this.getReactApplicationContext();
        // Environment.getExternalStoragePublicDirectory doesn't work as expected on Android Q
        // https://developer.android.com/reference/android/os/Environment#getExternalStoragePublicDirectory(java.lang.String)
        final File pubDirectory = context.getExternalFilesDir(Environment.DIRECTORY_DOWNLOADS);
        final File filename = new File(pubDirectory, exportDBFileName);

        return filename;
    }

    @ReactMethod
    public void exportUnencryptedDatabase(final String accountData, final String password, final Callback callback) {
        Log.d(TAG, "login");

        final File newFile = getExportDBFile();

        this.utils.migrateKeyStoreDir(accountData, password);
        String result = Statusgo.exportUnencryptedDatabase(accountData, password, newFile.getAbsolutePath());
        if (result.startsWith("{\"error\":\"\"")) {
            Log.d(TAG, "Login result: " + result);
        } else {
            Log.e(TAG, "Login failed: " + result);
        }
    }

    @ReactMethod
    public void importUnencryptedDatabase(final String accountData, final String password) {
        Log.d(TAG, "importUnencryptedDatabase");

        final File newFile = getExportDBFile();

        this.utils.migrateKeyStoreDir(accountData, password);
        String result = Statusgo.importUnencryptedDatabase(accountData, password, newFile.getAbsolutePath());
        if (result.startsWith("{\"error\":\"\"")) {
            Log.d(TAG, "import result: " + result);
        } else {
            Log.e(TAG, "import failed: " + result);
        }
    }

}

File: modules/react-native-status/android/src/main/java/im/status/ethereum/module/DatabaseManager.kt
package im.status.ethereum.module

import android.content.Context
import android.os.Environment
import android.util.Log
import com.facebook.react.bridge.Callback
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import statusgo.Statusgo
import java.io.File
import org.json.JSONObject
import org.json.JSONException

class DatabaseManager(private val reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {

    private val utils = Utils(reactContext)

    override fun getName() = "DatabaseManager"

    private fun getExportDBFile(): File {
        StatusBackendClient.getInstance()?.let {
            if (it.serverEnabled) {
                return File(it.rootDataDir, exportDBFileName)
            }
        }
        val pubDirectory = reactContext.getExternalFilesDir(Environment.DIRECTORY_DOWNLOADS)
        return File(pubDirectory, exportDBFileName)
    }

    @ReactMethod
    fun exportUnencryptedDatabase(accountData: String, password: String, callback: Callback) {
        Log.d(TAG, "exportUnencryptedDatabase")

        val newFile = getExportDBFile()

        utils.migrateKeyStoreDir(accountData, password)
        
        try {
            val accountJson = JSONObject(accountData)
            
            val params = JSONObject().apply {
                put("account", accountJson)
                put("password", password)
                put("databasePath", newFile.absolutePath)
            }
            
            val jsonParams = params.toString()
            StatusBackendClient.executeStatusGoRequestWithCallback(
                endpoint = "ExportUnencryptedDatabaseV2",
                requestBody = jsonParams,
                statusgoFunction = { Statusgo.exportUnencryptedDatabaseV2(jsonParams) },
                callback = null
            )
            callback.invoke(newFile.absolutePath)

        } catch (e: JSONException) {
            Log.e(TAG, "Error parsing account data: ${e.message}")
        }
    }

    @ReactMethod
    fun importUnencryptedDatabase(accountData: String, password: String) {
        Log.d(TAG, "importUnencryptedDatabase")

        val newFile = getExportDBFile()

        utils.migrateKeyStoreDir(accountData, password)

        try {
            val accountJson = JSONObject(accountData)
            
            val params = JSONObject().apply {
                put("account", accountJson)
                put("password", password)
                put("databasePath", newFile.absolutePath)
            }
            
            val jsonParams = params.toString()
            
            StatusBackendClient.executeStatusGoRequest(
                endpoint = "ImportUnencryptedDatabaseV2",
                requestBody = jsonParams,
                statusgoFunction = { Statusgo.importUnencryptedDatabaseV2(jsonParams) }
            )
        } catch (e: JSONException) {
            Log.e(TAG, "Error parsing account data: ${e.message}")
        }
    }

    companion object {
        private const val TAG = "DatabaseManager"
        private const val exportDBFileName = "export.db"
    }
}
